src/waveform/widgets/qopengl/waveformwidget.cpp

```diff
@@ -0,0 +1,61 @@
+#include "waveform/widgets/qopengl/waveformwidget.h"
+
+#include "waveform/renderers/qopengl/iwaveformrenderer.h"
+#include "widget/wwaveformviewer.h"
+
+using namespace qopengl;
+
+WaveformWidget::WaveformWidget(const QString& group, QWidget* parent)
+        : WGLWidget(parent), WaveformWidgetAbstract(group) {
+}
+
+WaveformWidget::~WaveformWidget() {
+    makeCurrentIfNeeded();
+    for (int i = 0; i < m_rendererStack.size(); ++i) {
+        delete m_rendererStack[i];
+    }
+    m_rendererStack.clear();
+    doneCurrent();
+}
+
+mixxx::Duration WaveformWidget::render() {
+    makeCurrentIfNeeded();
+    renderGL();
+    doneCurrent();
+    // not used, here for API compatibility
+    return mixxx::Duration();
```

Comment:
Looking at the other widgets, it "returns timer for painter setup", which is not relevant in this case. Also, this return value is not used at all, so it might be better to just remove it everywhere. But I will make the comment a bit more clear.